lndclient/versioner_client.go

```diff
@@ -0,0 +1,179 @@
+// lnd < 0.10.0 because any version previous to 0.10.0 doesn't have the version
+// endpoint implemented!
+func CheckVersionCompatibility(client VersionerClient, major, minor,
+	patch uint32, tags []string) error {
+
+	// First, test that the version RPC is even implemented.
+	version, err := client.GetVersion(context.Background())
+	if err != nil {
+		// The version service has only been added in lnd v0.10.0. If
+		// we get an unimplemented error, it means the lnd version is
+		// definitely older than that.
+		s, ok := status.FromError(err)
+		if ok && s.Code() == codes.Unimplemented {
+			return ErrVersionCheckNotImplemented
+		}
+		return fmt.Errorf("GetVersion error: %v", err)
+	}
+
+	// Now check the version and make sure all required build tags are set.
+	err = assertVersionCompatible(version, major, minor, patch)
+	if err != nil {
+		return err
+	}
+	err = assertBuildTagsEnabled(version, tags)
+	if err != nil {
+		return err
+	}
+
+	// All check positive, version is fully compatible.
+	return nil
+}
+
+// VersionString returns a nice, human readable string of a version returned by
+// the VersionerClient, including all build tags.
+func VersionString(version *verrpc.Version) string {
+	if version == nil {
+		return "<nil>"
+	}
+
+	short := VersionStringShort(version)
+	enabledTags := strings.Join(version.BuildTags, ",")
+	return fmt.Sprintf("%s, build tags '%s'", short, enabledTags)
+}
+
+// VersionStringShort returns a nice, human readable string of a version
+// returned by the VersionerClient.
+func VersionStringShort(version *verrpc.Version) string {
+	if version == nil {
+		return "<nil>"
+	}
+
+	versionStr := fmt.Sprintf(
+		"v%d.%d.%d", version.AppMajor, version.AppMinor,
+		version.AppPatch,
+	)
+	if version.AppPreRelease != "" {
+		versionStr = fmt.Sprintf(
+			"%s-%s", versionStr, version.AppPreRelease,
+		)
+	}
+	return versionStr
+}
+
+// assertVersionCompatible makes sure the detected lnd version is compatible
+// with our current version requirements.
+func assertVersionCompatible(actualVersion *verrpc.Version, expectedMajor,
+	expectedMinor, expectedPatch uint32) error {
+
+	// We need to check the versions parts sequentially as they are
+	// hierarchical.
+	if actualVersion.AppMajor != expectedMajor {
+		if actualVersion.AppMajor > expectedMajor {
+			return nil
+		}
+		return ErrVersionIncompatible
+	}
+
+	if actualVersion.AppMinor != expectedMinor {
+		if actualVersion.AppMinor > expectedMinor {
+			return nil
+		}
+		return ErrVersionIncompatible
+	}
+
+	if actualVersion.AppPatch != expectedPatch {
+		if actualVersion.AppPatch > expectedPatch {
+			return nil
+		}
+		return ErrVersionIncompatible
+	}
+
+	// The actual version and expected version are identical.
+	return nil
+}
+
+// assertBuildTagsEnabled makes sure all required build tags are set.
+func assertBuildTagsEnabled(actualVersion *verrpc.Version,
+	requiredTags []string) error {
+
+	enabled := strings.Join(actualVersion.BuildTags, ",")
```

Comment:
Probably cleaner not to use a string for this, but store the actual tags in a map.